beets/dbcore/query.py

```diff
@@ -1061,3 +1066,23 @@ def __eq__(self, other) -> bool:
 
     def __hash__(self) -> int:
         return 0
+
+
+class SmartArtistSort(FieldSort):
+    """Sort by artist (either album artist or track artist),
+    prioritizing the sort field over the raw field.
+    """
+
+    def order_clause(self):
+        order = "ASC" if self.ascending else "DESC"
+        collate = "COLLATE NOCASE" if self.case_insensitive else ""
+        field = self.field
+
+        return f"COALESCE(NULLIF({field}_sort, ''), {field}) {collate} {order}"
+
+    def sort(self, objs: list[AnyModel]) -> list[AnyModel]:
+        def key(o):
+            val = o[f"{self.field}_sort"] or o[self.field]
+            return val.lower() if self.case_insensitive else val
+
+        return sorted(objs, key=key, reverse=not self.ascending)
```

Comment:
Nice catch, making this general and independent of the specifics of beets' library items! Maybe, it should also be renamed to something more generic (`SmartSort`?) with a corresponding docstring (which could still mention the `artist`/`artist_sort` example)?

Reply:
Ah yeah, good point! 

Ah actually, if you look at `sort` implementation, you can see that we've got something hardcoded there that ties it to the `artist` field, as it firstly tries to sort by `{field}_sort` field.

Reply:
> Ah actually, if you look at `sort` implementation, you can see that we've got something hardcoded there that ties it to the `artist` field, as it firstly tries to sort by `{field}_sort` field.

Ok, but conceptually, this is a more generic mechanism now which could be used for other fields.

I think this boils down to deciding whether we treat this as generic dbcore code (in that case, name and docs should be adapted) or as `Library`-specific (in that case, this code should remain in `beets/library.py`). Currently, it's somewhere half-way, eroding the separation of concerns we have between `library` and `dbcore`.

Reply:
I think we actually agree here 😅

I don't have much of an opinion on where it belongs, I'd just argue for a rename and doc rewording in the case of moving it to `dbcore`.

Reply:
TBH, I don't have a really good idea, either...

Let's not block the PR over this, feel free to merge it as-is!